### ISSUE TITLE
TLS Validation Context (same config) not working on different versions

### ISSUE THREAD

**Issue by lalernehl** [envoyproxy/envoy#26788]
*Title*: *TLS Validation Context (same config) not working on different versions*

*Description*:
I have configured a cluster to do TLS verification and works fine in envoy v1.25.4 (config v3), however it fails running v1.23.1 (config v3) or v1.13.1 (config v2)

Same request for all 3 tests:
```
curl --location 'http://localhost:8080/' \
--header 'host: echo.staging.com' \
--header 'x-forwarded-proto: https'
```

Reponses:
```
v1.25.4
status code: 200

v1.23.1
status code: 503
upstream connect error or disconnect/reset before headers. reset reason: connection failure, transport failure reason: TLS error: 268435581:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED

v1.13.1
status code: 503
upstream connect error or disconnect/reset before headers. reset reason: connection failure
```

*Repro steps*:
Use the following `dockerfile` to run envoy (all 3 versions)
```
# FROM envoyproxy/envoy:v1.13.1
# COPY envoy1.13.1-V2.yaml /etc/envoy/envoy.yaml

# FROM envoyproxy/envoy:v1.25.4
# COPY envoy1.23.1-V3.yaml /etc/envoy/envoy.yaml

FROM envoyproxy/envoy:v1.23.1
COPY envoy1.23.1-V3.yaml /etc/envoy/envoy.yaml

COPY intermediateCAst3.pem /etc/envoy/intermediateCAst3.pem

CMD ["envoy", "-l", "trace", "-c", "/etc/envoy/envoy.yaml"]
```

*Admin and Stats Output*:
`admin/certs`
```
{
 "certificates": [
  {
   "ca_cert": [],
   "cert_chain": []
  },
  {
   "ca_cert": [
    {
     "path": "/etc/envoy/intermediateCAst3.pem",
     "serial_number": "3f7150e85fbd26a96d24345f4672c3e5998e400a",
     "subject_alt_names": [],
     "days_until_expiration": "1037",
     "valid_from": "2021-02-18T20:28:15Z",
     "expiration_time": "2026-02-17T20:28:45Z"
    }
   ],
   "cert_chain": []
  },
  {
   "ca_cert": [
    {
     "path": "/etc/envoy/intermediateCAst3.pem",
     "serial_number": "3f7150e85fbd26a96d24345f4672c3e5998e400a",
     "subject_alt_names": [],
     "days_until_expiration": "1037",
     "valid_from": "2021-02-18T20:28:15Z",
     "expiration_time": "2026-02-17T20:28:45Z"
    }
   ],
   "cert_chain": []
  }
 ]
}
```
Boring SSL configs:
```
"version": "edd69583372955fdfa0b8ca3820dd7312c094e46/1.23.1/Clean/RELEASE/BoringSSL",

"version": "13667370008884b39712b9d83ddb1d1626dbad32/1.25.4/Clean/RELEASE/BoringSSL",

"version": "b67c14052c49890a7e3afe614d50979c346c024b/1.13.1/Clean/RELEASE/BoringSSL",
```

Also upstream service cert has DNS setup
```
Extension=Subject Alternative Name ( 2.5.29.17 )
====Critical=NO
====DNS Name=echo.staging.com
```

*Config*:

Tested with v1.13.1 (does not work!)
```
admin:
  access_log_path: /dev/stdout
  address:
    socket_address:
      protocol: TCP
      address: 0.0.0.0
      port_value: 8888
static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 8080
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
          stat_prefix: ingress_http
          access_log:
            - name: envoy.file_access_log
              typed_config:
                "@type": type.googleapis.com/envoy.config.accesslog.v2.FileAccessLog
                path: /dev/stdout
          codec_type: AUTO
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match:
                  prefix: "/echo"
                request_headers_to_add:
                - header:
                    key: "X-Forwarded-Proto"
                    value: "https"
                  append: false
                route:
                  host_rewrite: echo.staging.com
                  cluster: echo_tls
          http_filters:
          - name: envoy.router
  clusters:
  - name: echo_tls
    connect_timeout: 2.5s
    type: LOGICAL_DNS
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: echo_tls
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: X.X.X.X
                port_value: YYYY
    transport_socket:
      name: envoy.transport_sockets.tls
      typed_config:
        "@type": type.googleapis.com/envoy.api.v2.auth.UpstreamTlsContext
        sni: echo.staging.com
        common_tls_context:
          validation_context:
            # match_subject_alt_names:
            # - exact: "echo.staging.com"
            verify_subject_alt_name:
              - echo.staging.com
            trusted_ca:
              filename: /etc/envoy/intermediateCAst3.pem
```

Tested with v1.25.4 (works!) and 1.23.1 (does not work!)
```
admin:
  access_log:
          - name: envoy.access_loggers.stdout
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.access_loggers.stream.v3.StdoutAccessLog
  address:
    socket_address:
      address: 0.0.0.0
      port_value: 8888

static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 8080
    filter_chains:
    - filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: ingress_http
          access_log:
          - name: envoy.access_loggers.stdout
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.access_loggers.stream.v3.StdoutAccessLog
          http_filters:
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match:
                  prefix: "/echo"
                request_headers_to_add:
                - header:
                    key: x-forwarded-proto
                    value: https
                  append: false
                route:
                  host_rewrite_literal: echo.staging.com
                  cluster: echo_tls
  clusters:
  - name: echo_tls
    type: LOGICAL_DNS
    dns_lookup_family: V4_ONLY
    load_assignment:
      cluster_name: echo_tls
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: X.X.X.X
                port_value: YYYY
    transport_socket:
      name: envoy.transport_sockets.tls
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
        sni: echo.staging.com
        common_tls_context:
          validation_context:
            match_typed_subject_alt_names:
            - san_type: DNS
              matcher:
                exact: "echo.staging.com"
            trust_chain_verification: VERIFY_TRUST_CHAIN
            trusted_ca:
              filename: /etc/envoy/intermediateCAst3.pem
```

*Logs*:

v1.23.1
```
...
[2023-04-17 14:55:03.428][17][debug][connection] [source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:207] X509_verify_cert: certificate verification error at depth 1: unable to get issuer certificate
[2023-04-17 14:55:03.429][17][trace][connection] [source/extensions/transport_sockets/tls/ssl_handshaker.cc:52] [C5] ssl error occurred while read: SSL
[2023-04-17 14:55:03.429][17][debug][connection] [source/extensions/transport_sockets/tls/ssl_socket.cc:228] [C5] remote address:10.102.220.103:23235,TLS error: 268435581:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
[2023-04-17 14:55:03.429][17][debug][connection] [source/common/network/connection_impl.cc:250] [C5] closing socket: 0
[2023-04-17 14:55:03.429][17][debug][connection] [source/extensions/transport_sockets/tls/ssl_socket.cc:228] [C5] remote address:10.102.220.103:23235,TLS error: 268435581:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
[2023-04-17 14:55:03.429][17][trace][connection] [source/common/network/connection_impl.cc:418] [C5] raising connection event 0
[2023-04-17 14:55:03.429][17][debug][client] [source/common/http/codec_client.cc:108] [C5] disconnect. resetting 0 pending requests
[2023-04-17 14:55:03.429][17][debug][pool] [source/common/conn_pool/conn_pool_base.cc:483] [C5] client disconnected, failure reason: TLS error: 268435581:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
[2023-04-17 14:55:03.429][17][debug][router] [source/common/router/router.cc:1198] [C3][S12284682140941904868] upstream reset: reset reason: connection failure, transport failure reason: TLS error: 268435581:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
[2023-04-17 14:55:03.429][17][trace][main] [source/common/event/dispatcher_impl.cc:250] item added to deferred deletion list (size=1)
[2023-04-17 14:55:03.429][17][debug][http] [source/common/http/filter_manager.cc:901] [C3][S12284682140941904868] Sending local reply with details upstream_reset_before_response_started{connection_failure,TLS_error:_268435581:SSL_routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED}
...
```
v1.13.1
```
...
[2023-04-17 14:15:11.340][18][debug][connection] [source/extensions/transport_sockets/tls/ssl_socket.cc:226] [C2] TLS error: 268435581:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
[2023-04-17 14:15:11.340][18][debug][connection] [source/common/network/connection_impl.cc:192] [C2] closing socket: 0
[2023-04-17 14:15:11.340][18][debug][client] [source/common/http/codec_client.cc:91] [C2] disconnect. resetting 0 pending requests
[2023-04-17 14:15:11.340][18][debug][pool] [source/common/http/http1/conn_pool.cc:136] [C2] client disconnected, failure reason: TLS error: 268435581:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
[2023-04-17 14:15:11.340][18][debug][pool] [source/common/http/http1/conn_pool.cc:166] [C2] purge pending, failure reason: TLS error: 268435581:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
[2023-04-17 14:15:11.340][18][debug][router] [source/common/router/router.cc:990] [C0][S13097465912368339894] upstream reset: reset reason connection failure
[2023-04-17 14:15:11.340][18][debug][http] [source/common/http/conn_manager_impl.cc:1417] [C0][S13097465912368339894] Sending local reply with details upstream_reset_before_response_started{connection failure,TLS error: 268435581:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED}
...
```

Please let me know if there is any other information I can send to troubleshoot.

Thank you in advance!

**Comment by htuch** [comment]
We don't support v2 xDS any more, and these versions are well outside a reasonable support window, so I'm going to close this one out. @ggreenway for visibility.

**Comment by kyessenov** [comment]
Can you post the successful trace log?
Boringssl ships cipher changes so maybe it's some default change when Envoy upgraded it.

**Comment by ggreenway** [comment]
Re-opening per #26816

**Comment by ggreenway** [comment]
I think the cause is that first log line: `[2023-04-17 14:55:03.428][17][debug][connection] [source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:207] X509_verify_cert: certificate verification error at depth 1: unable to get issuer certificate`. I can't tell why that's happening without seeing the full chain of the cert being verified and the CA.

It may be fixed due to #22350.

**Comment by lalernehl** [comment]
Yeah, we actually got around this problem by concatenating both root and intermediate into the same PEM
